tasks/network-interfaces.nix: Enable ip_forwarding for ipv4 and p…

kfiz opened the pull request:
…roxy_ndp for ipv6 for interfaces that set proxy_ARP

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
Fixes #62339

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/4

brainrape commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -963,11 +963,14 @@ in
       optionalString hasBonds "options bonding max_bonds=0";
 
     boot.kernel.sysctl = {
+      "net.ipv4.conf.all.forwarding" = mkDefault (any (i: i.proxyARP) interfaces);
```
This is a duplicate of line #968

brainrape commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -963,11 +963,14 @@ in
       "net.ipv6.conf.all.forwarding" = mkDefault (any (i: i.proxyARP) interfaces);
     } // listToAttrs (flip concatMap (filter (i: i.proxyARP) interfaces)
-        (i: flip map [ "4" "6" ] (v: nameValuePair "net.ipv${v}.conf.${i.name}.proxy_arp" true)))
+        (i: flip map [ "4" "6" ] (v: if v == "4" 
```
the `if` can be all the way inside, like `${if v == "4" then "arp" else "ndp"}`

kfiz replied:
Thanks for the comment. I simplified this.

brainrape commented:
I see this problem is still in master.

brainrape commented:
Please fix, rebase and LGTM, seems like an easy win.

kfiz commented:
Done.

brainrape commented:
Hey, I meant
```
  } // listToAttrs (flip concatMap (filter (i: i.proxyARP) interfaces)
        (i: forEach [ "4" "6" ] (v: nameValuePair "net.ipv${v}.conf.${replaceChars ["."] ["/"] i.name}.proxy_${if i == "4" then "arp" else "ndp"}" true)))
```
but maybe what you wrote is more readable.

I tested the PR, works fine, sets `proxy_arp` and `proxy_ndp` sysctls to `1`.

@flokli could you take a look? Thank you!

flokli commented:
@kfiz Please make this one single commit, and format the commit message properly. Also, the commit is missing a description on the *why*, as per `CONTRIBUTING.md`:

> In addition to writing properly formatted commit messages, it's important to include relevant information so other developers can later understand *why* a change was made. While this information usually can be found by digging code, mailing list/Discourse archives, pull request discussions or upstream changes, it may require a lot of work.

@brainrape I think the variant currently present in this PR is more readable.

kfiz commented:
@flokli Done.

flokli reviewed (changes requested):
Also note the commit author seems to be different.

flokli commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -1055,11 +1055,13 @@ in
     } // listToAttrs (flip concatMap (filter (i: i.proxyARP) interfaces)
-        (i: forEach [ "4" "6" ] (v: nameValuePair "net.ipv${v}.conf.${replaceChars ["."] ["/"] i.name}.proxy_arp" true)))
+        (i: [(nameValuePair "net.ipv4.conf.${replaceChars ["."] ["/"] i.name}.proxy_arp" true)
+             (nameValuePair "net.ipv6.conf.${replaceChars ["."] ["/"] i.name}.proxy_ndp" true)]))
```
It seems GitHub ate my comment.

This reads like we always set a wrong proxy_ndp sysctl (so it never worked), and even if we set it now, the kernel-provided ndp proxy infrastructure requires addition of each individual address to ndp proxy for (see https://github.com/DanielAdolfsson/ndppd/blob/master/README#L57)

It seems more useful to just remove the NDP-specific bits from the `proxyARP` documentation, and let users either use ndppd (which is packaged in NixOS), or set options themselves.

Luckily, as it never worked before, we don't break any existing configurations.

kfiz replied:
Hmm...you are right. Setting proxy_ndp alone is not enough to make this work. 

The question for me here would be whether a user who sets this knows that she has to additionally set something like `ip -6 neigh add proxy <ip> dev <if>` or not. If she is aware of this then this PR is aligned with her expectations, if not than obviously not. 

In any way, another way to go apart from your suggestion would be to generate a warning message that tells users that additional settings have to be specified for this to work and alternatively point to ndppd.

flokli replied:
`proxyARP` shouldn't enable proxy NDP - these are two different concepts, and we shouldn't mix them. Given it never worked, I don't see a problem with removing the (disfunctional and dangerous) proxy NDP part.

kfiz replied:
Agreed. We shouldn't mix the two. It's better to force users to make this decision consciously.

flokli replied:
Can you push a new commit changing this?

kfiz replied:
Changed this.

kfiz committed:
networking-interfaces.nix: remove  broken NDP bits from  proxyARP

The `networking.interfaces.<name?>.proxyARP` option previously mentioned it would also enable IPv6 forwarding and `proxy_ndp`.

However, the `proxy_ndp` option was never actually set (the non-existing `net.ipv6.conf.proxy_arp` sysctl was set
instead). In addition `proxy_ndp` also needs individual entries for each ip to proxy for.

Proxy ARP and Proxy NDP are two different concepts, and enabling the latter
should be a conscious decision.

This commit removes the broken NDP support, and disables explicitly
enabling IPv6 forwarding (which is the default in most cases anyways)

Fixes #62339.

flokli reviewed (changes requested):
The description for `networking.interfaces.<name?>.proxyARP` needs to be updated - it still refers to the NDP bits.

Also, please update the commit message to something like this:

```
tasks/networking-interfaces.nix: remove broken NDP bits from proxyARP option

The `networking.interfaces.<name?>.proxyARP` previously mentioned it would also enable IPv6 forwarding and `proxy_ndp`.

However, the `proxy_ndp` option was never actually set (as it did set the non-existing `net.ipv6.conf.proxy_arp` sysctl instead), and `proxy_ndp` also needs individual entries
for each ip to proxy for.

Proxy ARP and Proxy NDP are two different concepts, and enabling the latter
should be a conscious decision.

This commit removes the broken NDP support, and disables explicitly
enabling IPv6 forwarding (which is the default in most cases anyways)

Fixes #62339.
```

kfiz commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -1055,11 +1055,12 @@ in
+      "net.ipv4.conf.all.forwarding" = mkDefault (any (i: i.proxyARP) interfaces);
       "net.ipv6.conf.all.disable_ipv6" = mkDefault (!cfg.enableIPv6);
       "net.ipv6.conf.default.disable_ipv6" = mkDefault (!cfg.enableIPv6);
       "net.ipv6.conf.all.forwarding" = mkDefault (any (i: i.proxyARP) interfaces);
```
@flokli What do you think. Are we good to go?

flokli commented:
Thanks!
